### PR TITLE
congestion: advanced balanced workloads

### PR DESCRIPTION
**Pull request by jakmeier**
Add configuration parameters to the balanced workload to adjust:
- How deep each transaction should be. (number of hops)
- Different receipt sizes for different depth-levels.
- How wide each transaction should be. (how many children each parent receipt spwans)

Note that the workload is still balanced evenly across all shards.

Using these new parameters, I also add two new workloads to the test suite:
- Increasing Size: Start with small receipts locally but produce large outgoing receipts.
- Shard War: Each shard multiplies small receipts locally and then blast 64 larger receipts per tx to the receiving shard.

These two workloads produce >1GB large queue in some workloads.

### CODE REVIEW

**Comment by jakmeier** [comment]
Some results on the "shard wars" strategy:

(copied from above: *Shard War: Each shard multiplies small receipts locally and then blast 64 larger receipts per tx to the receiving shard.*)

Summary table: (Note large queue sizes)
```txt
STRATEGY              BURNT GAS  TRANSACTIONS FINISHED   MAX QUEUE LEN  MAX QUEUE SIZE  MAX QUEUE PGAS
No queues             3988 PGas                  24701           72587          2.7 GB             538
Global TX stop        3964 PGas                  25069           29763          1.4 GB             119
Simple backpressure     91 PGas                    530           10774        584.4 MB              36
New TX last           3973 PGas                  25339           26084          1.2 GB             106
Traffic Light         3988 PGas                  25367            9908        408.3 MB              61
```

- `Simple backpressure` is in a deadlock, as usual. 
- All other strategies finish around 25k transactions and burn around 4000 Pgas, so all of them are able to stay busy.
- `No queues` requires 2.7 GB for storing receipts on the worst shard.
- `Global TX stop` and `New TX last` use 1.4 and 1.2 GB respectively, probably more than we can handle.
- `Traffic light` requires "onyl" 408 MB (below the configured 500 MB soft max). It probably is in amber state on all shards most of the time since it has 61 > 50 Pgas queued.

It gets even more interesting when looking at Grafana. Both `Global TX stop` and `New TX last` have a periodic behaviour whereas `Traffic light` has not.

image

image

image

It seems the first two strategies accept way more work than they can handle and then once the receipts fan-out kicks in and there are 64x more receipts, they shut down new transactions completely and work until they have completed the accepted work. This behaviour on repeat gives the zig-zag pattern.

The traffic light strategy on the other hand has the benefit that it limits the accepted work based on attached gas. Thus it never accepts too much work at once and reaches a stable state quite quickly.

In conclusion: **Limiting inflow by attached gas seems like a good idea.**

**Comment by jakmeier** [comment]
And the same analysis for `IncreasingSize`.
(*Increasing Size: Start with small receipts locally but produce large outgoing receipts.*)

```txt
STRATEGY              BURNT GAS  TRANSACTIONS FINISHED   MAX QUEUE LEN  MAX QUEUE SIZE  MAX QUEUE PGAS
No queues             4045 PGas                  28526            3440          1.3 GB             246
Global TX stop        4070 PGas                  28575            4861          2.1 GB             214
Simple backpressure   4074 PGas                  28935            4512          2.0 GB             247
New TX last           4070 PGas                  29164            1254        456.1 MB              64
Traffic Light         4044 PGas                  28933             844        315.1 MB              56
```
First observations:
- Again, the number of transacations finished is very similar for all.
- `Simple backpressue` didn't deadlock this time.

Now it gets trickier.
- `No queues` leads to 1.3 GB usage while `Global TX stop` and `Simple backpressure` use more than 2 GB. Hence these two strategies perform worse than if we do nothing and keep today's behaviour!

This is odd but maybe we can makes sense of it with Grafana.

Starting with the "No queues" strategy, it simply accepts more and more transactions. It uses more and more memory but it has no weird behaviour.

image

Now look at "Global TX stop". We have a similar steady increase of pending transactions with an added oscillation. This comes in as we hit 1000 delayed receipts and limit new incoming transactions. This has the effect that the largest receipts which are at the end of each transaction get produced in a more concentrated way than if we just continued accepting more receipts.

If you look at the mean values for queue sizes, the required memory is actually smaller compared to `No queues`. But the peaks are worse.

image

It's the same story for `Simple backpressure` just with different conditions for when we start limiting new transactions. Mean values are even smaller but the peaks are about the same.

image



Now, if we look at `New TX last` in comparison, we can see the benefit of very aggressively avoiding new transactions. Thanks to only having ~400 transaction in-flight rather than ~2000, the required memory is much smaller.
There is still an oscillation but with a shorter period and lower peaks.

image

Finally, the `Traffic Light` has no oscillation to speak of. My interpretation is as follows: It fills the incoming buffers up to 50 Pgas the same way `No queue` does and then hits the amber state in all shards around the same time. Then, no new transactions are accepted for a moment. But as we start working down on the accumulated attached gas in the system, we slowly start accepting transactions again.
And this is the important part: Accepting new transactions is done more smoothly compared to the other strategies because it also accounts for attached gas.  The other strategies only look at number of delayed receipts. When we remove a large receipt and replace it with a small receipt, it can execute the new receipt quickly and will therefore start accepting more and more transactions in the following blocks. Up until the large receipts are kicking in again.
But the traffic light treats the new receipts already as expensive because of all the attached gas it has on it. It will therefore not overcommit to too many new transactions.

And then, thanks to releasing new transaction more smoothly, we avoid oscillation.

Of course, it also helps that the traffic light hits the limit of 50 Pgas at around 750 transaction in-flight, which is early enough that we don't use too much memory at any point.

image

In conclusion:
=> **Limiting by number of receipts can lead to undesirable oscillating behaviour.**
=> **Aggressively limiting incoming transactions works great to reduce the memory requirements.**

**Review by wacban — APPROVED** [review]
LGTM
